File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int result = 0;
    int[] d ={1, -1};
    int[] numbers_1;
    int target_1;
    int len;
    public int solution(int[] numbers, int target) {
        numbers_1 = numbers;
        target_1 = target;
        len = numbers.length;
        
        dfs(0, 0);
        
        return result;
    }
    public void dfs(int cnt, int num){
        if(cnt == len){
            if(num == target_1){
                result++;
            }
            return;
        }
        for(int i=0; i < 2; i++){
            dfs(cnt + 1, num + d[i] * numbers_1[cnt]);
        }
    }
    
    
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;


static int s_target, s_idx;
static int cnt = 0;
static vector<int> d{-1, 1};
static vector<int> s_numbers;

void dfs(int idx, int sum){
    if(idx == s_idx){
        if(sum == s_target) cnt ++;
        return;
    }
    
    for(int i = 0; i < 2; i ++){
        dfs(idx + 1, sum + (d[i] * s_numbers[idx]));
    }
}

int solution(vector<int> numbers, int target) {
    int answer = 0;
    s_target = target;
    s_idx = numbers.size();
    s_numbers = numbers;
    
    
    dfs(0, 0);
    answer = cnt;
    return answer;
}